src/lib/pubkey/pubkey.cpp

```diff
@@ -244,21 +240,36 @@ void check_der_format_supported(Signature_Format format, size_t parts) {
 
 }  // namespace
 
-PK_Signer::PK_Signer(const Private_Key& key,
-                     RandomNumberGenerator& rng,
-                     std::string_view emsa,
-                     Signature_Format format,
-                     std::string_view provider) {
-   m_op = key.create_signature_op(rng, emsa, provider);
+PK_Signer::PK_Signer(PK_Signature_Options options) {
+   // Appease GCC warning about a possibly dangling reference because when using
+   // std::reference_wrapper's implicit conversion to the contained reference
+   // the wrapper does not outlive the reference which seems to confuse GCC.
+   const auto key_wrapped = options.private_key().required();
+   const Private_Key& key = key_wrapped.get();
+
+   // TODO: The downstream algorithms should decide whether to require an RNG
+   //       (e.g. explicitly 'deterministic' signatures should not require it).
+   RandomNumberGenerator& rng = options.rng().required();
```

Comment:
Currently this enforces that all algorithms specify an RNG (via `with_rng()`), even if the algorithm doesn't need an RNG for signing, like some ECC-based schemes or hardware-backed adapters, or when the algorithm is configured to work deterministically (via `with_deterministic_signature()`). In those cases, it shouldn't be required to pass an RNG, and perhaps passing one should even produce an error (because the RNG wasn't consumed).